src/auth/AuthContext.tsx

```diff
@@ -0,0 +1,175 @@
+  user: FirebaseAuthTypes.User | null;
+};
+
+type AuthReducerAction =
+  | {
+      type: 'SET_USER';
+      user: FirebaseAuthTypes.User | null;
+    }
+  | {type: 'SET_ABT_CUSTOMER_ID'; abtCustomerId: string | undefined};
+
+type AuthReducer = (
+  prevState: AuthReducerState,
+  action: AuthReducerAction,
+) => AuthReducerState;
+
+const authReducer: AuthReducer = (prevState, action): AuthReducerState => {
+  switch (action.type) {
+    case 'SET_USER': {
+      return {
+        ...prevState,
+        isInitialized: true,
+        user: action.user,
+      };
+    }
+    case 'SET_ABT_CUSTOMER_ID': {
+      return {
+        ...prevState,
+        abtCustomerId: action.abtCustomerId,
+      };
+    }
+  }
+};
+
+const initialReducerState: AuthReducerState = {
+  isInitialized: false,
+  abtCustomerId: undefined,
+  user: null,
+};
+
+type AuthContextState = {
+  signInAnonymously: () => Promise<void>;
+  signOut: () => Promise<void>;
+  updateEmail: (email: string) => Promise<void>;
+  syncAbtCustomer: () => Promise<void>;
+} & AuthReducerState;
+
+const AuthContext = createContext<AuthContextState | undefined>(undefined);
+
+export default function AuthContextProvider({children}: PropsWithChildren<{}>) {
+  const [state, dispatch] = useReducer(authReducer, initialReducerState);
+  const {
+    preferences: {language},
+  } = usePreferences();
+
+  useEffect(() => {
+    if (language) auth().setLanguageCode(language);
+  }, [language]);
+
+  const onUserChanged = useCallback(
+    async function (user: FirebaseAuthTypes.User | null) {
+      dispatch({type: 'SET_USER', user});
+    },
+    [dispatch],
+  );
+
+  useEffect(() => {
+    syncAbtCustomer();
+  }, [state.user]);
+
+  const getAbtCustomerId = useCallback(
+    async function () {
+      if (!state.user) return undefined;
+      const idToken = await state.user.getIdTokenResult(true);
+      return idToken.claims['abt_id'] as string | undefined;
+    },
+    [state.user],
+  );
+
+  const syncAbtCustomer = useCallback(
+    async function () {
+      async function sync(retries: number = 0): Promise<string | undefined> {
+        if (retries > 3) return undefined;
+
+        const abtCustomerId = await getAbtCustomerId();
+        if (abtCustomerId) {
+          return abtCustomerId;
+        } else {
+          return sync(retries + 1);
+        }
+      }
+
+      const abtCustomerId = await sync();
+      dispatch({type: 'SET_ABT_CUSTOMER_ID', abtCustomerId});
+    },
+    [getAbtCustomerId],
+  );
+
+  useEffect(() => {
+    const subscriber = auth().onUserChanged(onUserChanged);
+    return subscriber;
+  }, [onUserChanged]);
+
+  const signInAnonymously = useCallback(async function () {
+    await auth().signInAnonymously();
+  }, []);
+
+  const signOut = useCallback(async function () {
+    await auth().signOut();
+  }, []);
+
+  useEffect(() => {
+    if (state.isInitialized && !state.user) {
+      signInAnonymously();
+    }
+  }, [state.isInitialized]);
```

Comment:
I've ran and tested this locally now and all seem good! Really nice including the log in/out buttons for local testing.

With that said I could not get a grasp on the flow of the login process just by reading this file. I had to start debugging and printing text to understand what was really happening. I think a couple comments and some small rearrangements could make the flow a little more clear for future maintainers.

Something like this?
```javascript
  // Subscribe to user changes. Will fire a onChangeEvent immediately on subscription.
  useEffect(
    () => auth().onUserChanged((user) => dispatch({type: 'SET_USER', user})),
    [],
  );

  // Sign in if the onChangeEvent fired immediately on subscription did not include user data.
  useEffect(() => {
    if (state.isInitialized && !state.user) {
      signInAnonymously();
    }
  }, [state.isInitialized]);
  ```

Reply:
Good idea, can add those!